### PR TITLE
Event Parameters Components.

### PR DESCRIPTION
**Pull request by jessetilro**
Implemented and tested the basic components needed for making Events parameterized. Merging into another branch in order to do a temporary code review of the components before integrating them with the rest of the components.

Contributes to resolving Issue #190.
Tested to cover 100% of the lines.
All static errors fixed.

### CODE REVIEW

**Comment by dalderliesten** [comment]
Looks good to me.

File: src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ * (EventParameterValue) or a continuous one (Integer).
+ * 
+ * @author Jesse Tilro
+ *
+ */
+public class EventParameter {
+
+  private String key;
+  private EventParameterValues value;
+  private int number;
```
Comment (FlorisXll) [inline src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java]: Perhaps call it numberValue or valueNumber?

**Comment by dmalarev** [comment]
Looks good, though i believe the classes EventParameter and EventParameters might become confusing due to their name being only different by one letter

File: src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java [hunk 2]
```diff
@@ -56,17 +56,67 @@ public String getKey() {
...
+   * Generate a Hash Code for this instance.
+   * 
+   * @return A likely unique hash code for this instance.
    */
-  public boolean isDiscrete() {
-    return !isContinuous();
+  @Override
+  public int hashCode() {
+    int result = 42;
+    int prime = 37;
```
Comment (FlorisXll) [inline src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java]: I would include this in the constants class.

**Comment by jessetilro** [comment]
Dmitry you are right, maybe I should rename it to EventParameterCollection?

File: src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java (hunk 1 above)
Comment (jessetilro) [inline src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java]: Since `value` and `number` are practically alternatives and not complementary, I believe their names should be fine like this, right?

File: src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java (hunk 2 above)
Comment (jessetilro) [inline src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java]: No they are class scoped, not application scoped. Each hashCode method may use different numbers for these purposes.

File: src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java (hunk 1 above)
Comment (wouterz) [inline src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java]: Only remark I have is that you have these two paremters, number and value. But then in the constructor you assign this.number = value

`  public EventParameter(String key, int value) {
     this.key = key;
    this.value = EventParameterValues.GENERIC;
    this.number = value;
  }` while correct this looks confusing. I suggest replacing this by this.number = number

File: src/test/java/nl/tudelft/pixelperfect/event/EventParametersTest.java [hunk 3]
```diff
@@ -111,6 +122,14 @@ public void testEqualsDifferentSizes() {
   }
 
   /**
+   * An EventParameters instance should no be considered equal to something of another type.
```
Comment (FlorisXll) [inline src/test/java/nl/tudelft/pixelperfect/event/EventParametersTest.java]: You want to test whether something is not equal to another type, but you only test null, which would never be true. Maybe just test it against a random other class.

File: src/test/java/nl/tudelft/pixelperfect/event/EventParameterTest.java [hunk 4]
```diff
@@ -98,6 +117,14 @@ public void testEqualsDifferentNumbers() {
   }
 
   /**
+   * An EventParameter instance should no be considered equal to something of another type.
```
Comment (FlorisXll) [inline src/test/java/nl/tudelft/pixelperfect/event/EventParameterTest.java]: See my comment on EventParametersTest line 125

**Comment by wouterz** [comment]
LGTM

**Comment by dmalarev** [comment]
@jessetilro EventParameterCollection sounds good

File: src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java (hunk 1 above)
Comment (FlorisXll) [inline src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java]: Yes, but the naming is rather useful for future methods. For example, what if I am in a method and it just says 'number', I would have to backtrack (ofc, that would not be hard here).

It is just a little bit of clarification of an important variable, it is not a fatal error.

Comment (dalderliesten) [inline src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java]: I agree with Floris, it's not a huge deal right now, but try to always give it a descriptor and an identifier.  Perhaps:  "parameterNumber" or "thisValCountsInts."

File: src/test/java/nl/tudelft/pixelperfect/event/EventParametersTest.java (hunk 3 above)
Comment (jessetilro) [inline src/test/java/nl/tudelft/pixelperfect/event/EventParametersTest.java]: You're right, but null and a different type are handled exactly the same way: the instanceof yields false. It is just for the line coverage, so it really doesn't matter whether it is null or an instance of another class.

File: src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java (hunk 1 above)
Comment (jessetilro) [inline src/main/java/nl/tudelft/pixelperfect/event/EventParameter.java]: Why make attribute names so complicated? You never use them when they are well encapsulated (like here), you are only using the class's public interface. In addition there is a lot of javadoc surrounding the attribute and the methods it it involved in, so regardless of the name it should be perfectly clear what it stands for. Nonetheless I will take your comments into account and refactor it to something like number value, thanks.

File: src/test/java/nl/tudelft/pixelperfect/event/EventParametersTest.java (hunk 3 above)
Comment (FlorisXll) [inline src/test/java/nl/tudelft/pixelperfect/event/EventParametersTest.java]: Yes, I understand, that's why you get a hundred percent. But it's about the quality of the test instead coverage, right? :)

Comment (jessetilro) [inline src/test/java/nl/tudelft/pixelperfect/event/EventParametersTest.java]: Not in my opinion. Tests at this level are just to meet coverage guidelines, they hardly test any functionality. Nevertheless, I wil make an additional test case for an instance of another class.

**Comment by FlorisXll** [comment]
Look pretty good now, merging after Travis report.

**Comment by dalderliesten** [comment]
Much better, great work.
